### ISSUE TITLE
TestCoroutineScheduler leaks canceled delayed jobs

### ISSUE THREAD

**Issue by pyricau** [Kotlin/kotlinx.coroutines#3398]
Jobs that are cancelled are expected to become garbage collectable. Unfortunately, TestCoroutineScheduler does not listen to any job cancellation signal. TestCoroutineScheduler keeps cancelled jobs in its queue (`TestCoroutineScheduler#events`) until the current time of TestCoroutineScheduler is advanced past the original planned time of cancelled job.

I'm providing a unit test that reproduces this issue here: (TODO insert link)

This is unfortunately a major bug for us at Square / Block, because:
- TestCoroutineScheduler is used by Compose in its provided test rules
- Compose clickable element in scrollable containers set their pressed state using a delayed cancellable job ([source](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/Clickable.kt;l=437-447;drc=f7a365acb5caa222c5421eef6d39abfae56917d8)) and that delayed cancellable job has a transitive strong reference to the backing AndroidComposeView.
- As a result, if said AndroidComposeView gets detached, the TestCoroutineScheduler will prevent it from being garbage collected until its time is past the clickable timeout. In our UI tests, automatic leak detection triggers in between (as the view gets detached) and fails the test.
- So basically we ended up having to disable leak detection in all our Compose tests due to this TestCoroutineScheduler bug.

If we reach into `TestCoroutineScheduler#events` using reflection to clear all canceled jobs from the queue, the jobs do become garbage collectable and we can keep leak detection enabled in our Compose test (which is a huge deal for us)

**Comment by qwwdfsad** [comment]
Thanks for the detailed report!
@dkhalanskyjb who is the main maintainer of the test module is on vacation right now, so delayed (he-he) answer here is expected

**Comment by dkhalanskyjb** [comment]
Thank you! Yes, this is a bug in the test module. The fix, as is common, a single line of code.

**Comment by pyricau** [comment]
Nice fix, thanks!
